File: erpnext/production/report/production_and_disposal_target_report/production_and_disposal_target_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Production and Disposal Target Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": ("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"reqd": 1,
		}, 
		{
			"fieldname": "uinput",
			"label": ("Options"),
			"fieldtype": "Select",
			"options": ["Production", "Disposal"],
			"reqd" : 1,
		},
		{
			"fieldname": "cost_center",
			"label": ("Parent Branch"),
			"fieldtype": "Link",
			"options": "Cost Center",
			get_query: () => {
				var company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						disabled: 0,
						company: company,
						is_group: 1
					}
				};
			},
			"on_change": function(query_report) {
				var cost_center = frappe.query_report.get_filter_value("cost_center");
				query_report.trigger_refresh();
				if (cost_center) {
					frappe.call({
						method: "erpnext.custom_utils.get_branch_from_cost_center",
						args: {
							"cost_center": cost_center,
						},
						callback: function(r) {
							query_report.set_filter_value("branch", r.message);
							query_report.trigger_refresh();
						}
					})
				}
			},
			"reqd": 1,
		},
		{
			"fieldname": "branch",
			"label": ("Branch"),
			"fieldtype": "Link",
			"options": "Cost Center",
			// "read_only": 1,
			// "get_query": function() {
			//         var company = frappe.query_report.filters_by_name.company.get_value();
			//         return {"doctype": "Branch", "filters": {"company": company, "is_disabled": 0}}
			// }
			get_query: () => {
				var cost_center = frappe.query_report.get_filter_value("cost_center")
				var company = frappe.query_report.get_filter_value("company")
				if(cost_center!= 'Natural Resource Development Corporation Ltd - NRDCL')
				{
					return {
						filters: {
							disabled: 0,
							company: company,
							parent_cost_center: cost_center
						}
					};
				}
				else
				{
					return {
						filters: {
							disabled: 0,
							company: company,
							is_group: 0
						}
					};
				}
			}
		},
		{
			"fieldname": "location",
			"label": ("Location"),
			"fieldtype": "Link",
			"options": "Location",
			"get_query": function() {
				var branch = frappe.query_report.filters_by_name.branch.get_value();
				branch = branch.replace(' - NRDCL','');
				return {
					filters: {
						is_disabled: 0,
						branch: branch,
					}
				};
			}
		},
		{
			"fieldname": "item_group",
			"label": __("Material Group"),
			"fieldtype": "Link",
			"options": "Item Group",
			// "reqd": 1,
		},
                {
			"fieldname": "uom",
			"label": __("UOM"),
			"fieldtype": "Link",
			"options": "UOM",
			// "reqd": 1,
		},
	]
}

File: erpnext/production/report/production_and_disposal_target_report/production_and_disposal_target_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from erpnext.accounts.utils import get_child_cost_centers, get_period_date


def execute(filters=None):
       
	columns = get_columns(filters)
	data = get_data(filters)

	return columns, data

def get_columns(filters):
	if filters.uinput == "Production":
		cols= [
			("Region") + ":Link/Cost Center:140",
			("Branch") + ":Link/Branch:140",
			("Production Group") + ":Data:140",
			("Material Group") + ":Link/Item Group:140",
			("UOM") + ":Link/UOM:70",
			("1st Quarter") + ":Float:90",
			("2nd Quarter") + ":Float:90",
			("3rd Quarter") + ":Float:90",
			("4th Quarter") + ":Float:90",
			("Target Qty") + ":Float:90"
        	]
	else:
		cols=[
			("Region") + ":Link/Cost Center:140",
			("Branch") + ":Link/Branch:140",
			("Disposal Group") + ":Data:140",
			("Material Group") + ":Link/Item Group:140",
			("UOM") + ":Link/UOM:70",
			("1st Quarter") + ":Float:90",
			("2nd Quarter") + ":Float:90",
			("3rd Quarter") + ":Float:90",
			("4th Quarter") + ":Float:90",
			("Disposal Qty") + ":Data:97"
		]
	return cols 

def get_data(filters):
	all_ccs = get_child_cost_centers(filters.cost_center)

	if not filters.uinput:
		return []
	query = "select (select cc.parent_cost_center from `tabCost Center` cc where cc.name = pt.cost_center) as region, pt.branch, pdi.production_group, pt.item_group, pdi.uom, sum(pdi.quarter1), sum(pdi.quarter2), sum(pdi.quarter3), sum(pdi.quarter4), sum(pdi.quantity) from `tab{0} Target Item` pdi, `tabProduction Target` pt where pt.name = pdi.parent".format(filters.uinput) 

	#--------CONDITIONS------------------------------------------------------#
	if filters.branch:
		branch = str(filters.branch)
		branch = branch.replace(' - NRDCL','')
		query += " and pt.branch = '"+branch+"'"

	else:	
		query += " and pt.cost_center in {0} ".format(tuple(all_ccs))
			
	if filters.get("location"):
		query += " and pt.location = \'" + str(filters.location) + "\'"

	if filters.get("fiscal_year"):
		query += " and pt.fiscal_year = \'" + str(filters.fiscal_year) + "\'"

	if filters.get("item_group"):
		query += " and pt.item_group = '{0}'".format(filters.get("item_group"))
	if filters.get("uom"):
		query += " and pdi.uom = '{0}'".format(filters.get("uom"))

	query += " group by pdi.production_group " 
	# if frappe.session.user == "Administrator":
	# 	frappe.throw(query)
	return frappe.db.sql(query)
